docs/design/backups.md

```diff
@@ -0,0 +1,83 @@
+# Backup & Restore
+
+In particularly disastrous situations, we would like to be able to revert the etcd cluster to a previously known healthy state.
+This document lays out the design of how we want the healthy state to be captured, securely stored & safely restored to operator-managed etcd clusters.
+
+## Backup Resource
+
+We will introduce a new custom resource: `EtcdBackup`.
+
+```yaml
+apiVersion: etcd.improbable.io/v1alpha1
+kind: EtcdBackup
+metadata:
+  name: my-cluster-backup
+  namespace: default
+spec:
+  type: snapshot
+  destination:
+    gcsBucket:
+      bucketName: my-cluster-backup-bucket
+      creds:
+        secretRef: 
+          name: my-cluster-backup-bucket-credentials
+          namespace: default
+```
+
+This resource specifies that we desire a backup to be immediately taken as a [snapshot](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/recovery.md#snapshotting-the-keyspace) from the etcd API.
+This snapshot will be saved to a GCS bucket, accessed with credentials read from a given secret.
+The file would be named with a timestamp when the backup was started, for uniqueness.
+Backups will be taken by the controller, rather than a Job or Pod acting on behalf of the controller.
+
+Backup types could include
+  * `snapshot`: a `snapshot.db` file taken from the etcd API
+  * `keys`: a zip file containing all key-values as files
+  * `keys-json` a json file containing all key-values as a JSON file
+
+Only `snapshot` will be initially implemented.
+
+Destinations culd include
+  * `socket`: data would be sent to a TCP socket
+  * `gcsBucket`: pushed to a a Google Cloud Storage bucket
+  * `s3Bucket`: similar to gcsBucket, but for AWS.
+  * ... other hosted blob storage services
+
+Only `gcsBucket` will be initially implemented.
+
+A controller will be watching for these resources.
+When an `EtcdBackup` resource is deployed the controller will start taking the backup in the specified way, publishing an event to notify that the backup has started. 
```

Comment:
What happens if the backup cannot start due to the cluster being unavailable? Do we have some way to denote a backup's status? i.e. NotStarted/Pending/InProgress/Complete/Failed?

Reply:
I was expecting that after the backup is attempted the backup resource's status wil mark it as success or fail. Very much like a pod, I wouldn't exect an individual backup to retry but I would expect the scheduled backup controller to retry another backup. As an administrator i'd like to be able to list etcd backup resources and look for failures (or alert on them)